source/slang/slang-ir-metal-legalize.cpp

```diff
@@ -10,709 +10,1371 @@
-                UInt index = 0;
-                val = builder.emitSwizzle(fromVector->getElementType(), val, 1, &index);
+                if (fromBasicType->getOp() == kIROp_VoidType)
+                    return nullptr;
+                if (!as<IRBasicType>(toType))
+                    return nullptr;
                 if (toType->getOp() == kIROp_VoidType)
                     return nullptr;
             }
-        }
-        else if (auto fromBasicType = as<IRBasicType>(fromType))
-        {
-            if (fromBasicType->getOp() == kIROp_VoidType)
-                return nullptr;
-            if (!as<IRBasicType>(toType))
-                return nullptr;
-            if (toType->getOp() == kIROp_VoidType)
+            else
+            {
                 return nullptr;
+            }
+            return builder.emitCast(toType, val);
         }
-        else
-        {
-            return nullptr;
-        }
-        return builder.emitCast(toType, val);
-    }
-
-    void legalizeSystemValueParameters(EntryPointInfo entryPoint, DiagnosticSink* sink)
-    {
-        SLANG_UNUSED(sink);
 
         struct SystemValLegalizationWorkItem
         {
-            IRParam* param;
+            IRInst* var;
             String attrName;
             UInt attrIndex;
         };
-        List<SystemValLegalizationWorkItem> systemValWorkItems;
-
-        IRBuilder builder(entryPoint.entryPointFunc);
 
-        for (auto param : entryPoint.entryPointFunc->getParams())
+        std::optional<SystemValLegalizationWorkItem> tryToMakeSystemValWorkItem(IRInst* var)
         {
-            if (auto semanticDecoration = param->findDecoration<IRSemanticDecoration>())
+            if (auto semanticDecoration = var->findDecoration<IRSemanticDecoration>())
             {
                 if (semanticDecoration->getSemanticName().startsWithCaseInsensitive(toSlice("sv_")))
                 {
-                    systemValWorkItems.add({ param, String(semanticDecoration->getSemanticName()).toLower(), (UInt)semanticDecoration->getSemanticIndex() });
-                    continue;
+                    return { { var, String(semanticDecoration->getSemanticName()).toLower(), (UInt)semanticDecoration->getSemanticIndex() } };
                 }
             }
 
-            auto layoutDecor = param->findDecoration<IRLayoutDecoration>();
+            auto layoutDecor = var->findDecoration<IRLayoutDecoration>();
             if (!layoutDecor)
-                continue;
+                return {};
             auto sysValAttr = layoutDecor->findAttr<IRSystemValueSemanticAttr>();
             if (!sysValAttr)
-                continue;
+                return {};
             auto semanticName = String(sysValAttr->getName());
             auto sysAttrIndex = sysValAttr->getIndex();
-            systemValWorkItems.add({ param, semanticName, sysAttrIndex });
+
+            return std::make_optional<SystemValLegalizationWorkItem>({ var, semanticName, sysAttrIndex });
```

Comment:
Should be able to use the same initializer list style return here as the earlier return statements.